### PR TITLE
osd: async recovery

### PR DESCRIPTION
**Pull request by wonzhq**
Follow-on of https://www.spinics.net/lists/ceph-devel/msg33363.html

### CODE REVIEW

**Comment by wonzhq** [comment]
@athanatos @liewegas, need you to take a look, thx

**Comment by dachary** [comment]
jenkins test this please (eio ignored in master logs)

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -572,21 +572,42 @@ bool ReplicatedPG::is_degraded_or_backfilling_object(const hobject_t& soid)
...
-    if (is_backfill_targets(peer) &&
-	cmp(peer_info[peer].last_backfill, soid, get_sort_bitwise()) <= 0 &&
+  // Object is degraded if after last_backfill AND
+  // we are backfilling it
+  for (set<pg_shard_t>::iterator p = backfill_targets.begin();
+       p != backfill_targets.end();
+       ++p) {
+    if (*p == get_primary()) continue;
+    pg_shard_t backfill_peer = *p;
+    if (cmp(peer_info[backfill_peer].last_backfill, soid, get_sort_bitwise()) <= 0 &&
```
Comment (Jerry7X) [inline src/osd/ReplicatedPG.cc]: right , otherwise how to keep order for recovery and client io (which is different priority)

Comment (Jerry7X) [inline src/osd/ReplicatedPG.cc]: if act like backfill, we should not take it as auth shard when find_best_info.

**Comment by athanatos** [comment]
I won't have time to look at this in detail until Kraken is out.

File: src/osd/ReplicatedPG.cc (hunk 1 above)
Comment (wonzhq) [inline src/osd/ReplicatedPG.cc]: This is not true. Async recovery allows IOs on objects which are degraded on the async recovery targets. But if an object is degraded on the acting set OSDs, we still need to recover it first before handling IO.

**Comment by wonzhq** [comment]
Sure @athanatos.
Btw, this passes our internal teuthology testing on jewel. Hopefully not too much trouble on the current master..

**Comment by tchaikov** [comment]
@jdugin is also working on async recovery, see https://github.com/jdurgin/ceph/commits/wip-async-recovery

**Comment by jdurgin** [comment]
there's been a bit of convergent evolution on this - I'm working on a similar optimization that uses the normal log-based recovery process instead of backfill, which will generally be much faster, and applies to ec pools as well. I hope to have it ready in the next couple weeks.

Your approach here makes sense in general, but I think using backfill alone isn't worth the extra complexity - you can get a similar (though not quite the same) effect by turning down the pg log sizes to force backfill more often.

**Comment by wonzhq** [comment]
@jdurgin, not sure if I was saying it wrong in the mail list or somewhere, but in this PR I'm not trading  recovery for backfill. I'm still doing normal log-based recovery for the PG shards whose pglog overlaps with the authoritative shard and its content is far behind current.

**Comment by Peerless-Jing** [comment]
have you ever run fio+librbd to test the performance of random write? I came across an error when running this, this error is like "rbd_open failed. fio_rbd_connect failed."  
Anyway, could you please offer me a way to test the performance of random write? thx a lot! Looking forward to your replying!!

**Comment by liewegas** [comment]
We've merged an alternate approach

**Pull request closed without merging**